src/mm-input-mask/mm-input-mask.js

```diff
@@ -0,0 +1,396 @@
+				this.arimoLoaded = true;
+				this.maskConfig.filter(function(group) {
+					return group.type === _types.GROUP;
+				}).forEach(function(group) {
+					group.loaded = true;
+				},this);
+			}.bind(this));
+		},
+
+		ready: function() {
+
+			this.super();
+			if (!this.rules) {
+				this.rules = _rules;
+			}
+			if(!this.restrict) {
+				this.restrict = _restrict;
+			}
+			//TODO: remove this in favor of dom notifier
+			this.parseMask();
+
+		},
+
+		domReady: function() {
+			this.updateGroups();
+		},
+
+		attached: function() {
+			this.$.input.addEventListener("focus", this.handleFocus.bind(this));
+			this.$.input.addEventListener("blur", this.handleBlur.bind(this));
+		},
+
+		detached: function() {
+			this.$.input.removeEventListener("focus");
+			this.$.input.removeEventListener("blur");
+		},
+
+		parseMask: function() {
+			this.sepsLen = 0; //used to check 'empty' state for placeholder styling
+			var nodes = DataUtils.clone( this.$.content.getDistributedNodes() ),
+				groups = [],
+				seps = [];
+			this.maskConfig = nodes.map(function(node, index) {
+				switch(node.nodeName) {
+					case "GROUP":
+						var rng = _groupRange(node.getAttribute("size")),
+							rule = this.rules[node.getAttribute("rule")],
+							restrict = this.restrict[node.getAttribute("restrict")],
+							auto = node.getAttribute("autofill"),
+							args = node.textContent.trim(),
+							style = {
+								width:"auto"
+							};
+						if (restrict !== _restrict.numeric) style = {};
+
+						var o = {
+							index:index,
+							max:rng[1],
+							min:rng[0], 
+							rule:rule, 
+							restrict:restrict, 
+							args:args, 
+							auto:auto,
+							type:_types.GROUP, 
+							style:style,
+							loaded: false,
+							contentWidth:this.contentWidthHandler
+						};
+						groups.push(o);
+						return o;
+
+					case "SEP":
+						var chars = node.getAttribute("chars");
+						seps.push(chars);
+						this.sepsLen += chars.length;
+						return {
+							type:_types.SEP, 
+							value:chars, 
+							style:{
+								sep:true, 
+								placeholder:true
+							}
+						};
+				}
+			}, this);
+			this.seps = seps;
+			this.groups = groups;
+
+		},
+
+		contentWidthHandler: function(style, val, placeholder, loaded) {
+			var w = 0.0;
+			var check = val || placeholder || "";
+			while(check.length < this.max) {
+				check += "S";
+			}
+			if (!val && placeholder) {
+				w = Measure.textWidth(null, check, "italic 13px Arimo");
+			} else {
+				w = Measure.textWidth(null, check, "13px Arimo");
+			}
+			style.width = w + "px";
+			return style;
+		},
+
+		placeholderChanged: function(oldPlaceholder, newPlaceholder) {
+			this.chunkValue(this.placeholder, "placeholder");
+		},
+
+		valueChanged: function(oldValue, newValue) {
+			if (newValue) {
+				var placeholder = newValue.length <= this.sepsLen;
+				this.maskConfig.filter(function(m) { return m.type === _types.SEP}).forEach(function(sep) {
```

Comment:
Similar above: Why not .forEach on this.seps instead of a filtered subset of this.maskConfig?  Are you expecting them to be out of sync at some point?